crates/circuit/src/operations.rs

```diff
@@ -954,7 +959,41 @@ impl Operation for StandardGate {
                 )
             }),
             Self::CRXGate | Self::CRYGate | Self::CRZGate => todo!(),
-            Self::RGate => todo!(),
+            Self::RGate => Python::with_gil(|py| -> Option<CircuitData> {
+                let theta_expr = match &params[0] {
+                    Param::Float(theta) => Param::Float(*theta),
+                    Param::ParameterExpression(theta) => {
+                        Param::ParameterExpression(theta.clone_ref(py))
+                    }
+                    Param::Obj(_) => unreachable!(),
+                };
+                let (phi_expr1, phi_expr2) = match &params[0] {
+                    Param::Float(phi) => (Param::Float(*phi - 1.0), Param::Float(-*phi + 1.0)),
+                    Param::ParameterExpression(phi) => {
+                        let phiexpr1 = phi
+                            .call_method1(py, intern!(py, "__add__"), ((-PI / 2.0),))
+                            .expect("Unexpected Qiskit python bug");
+                        let phiexpr2 = phiexpr1
+                            .call_method1(py, intern!(py, "__rmul__"), (-1.0,))
+                            .expect("Unexpected Qiskit python bug");
+                        (
+                            Param::ParameterExpression(phiexpr1),
+                            Param::ParameterExpression(phiexpr2),
+                        )
+                    }
+                    Param::Obj(_) => unreachable!(),
+                };
+                let defparams = smallvec![theta_expr, phi_expr1, phi_expr2];
+                Some(
+                    CircuitData::from_standard_gates(
+                        py,
+                        1,
+                        [(Self::UGate, defparams, smallvec![Qubit(0)])],
+                        FLOAT_ZERO,
+                    )
+                    .expect("Unexpected Qiskit python bug"),
+                )
+            }),
```

Comment:
I added a `multiply_param` function in and a `sum_params` function in #12659 ([here](https://github.com/Qiskit/qiskit/pull/12659/files#diff-82a880e0be3e8b32095f0ce8d1c1ea09978d09013fca3b015d8210958b938fd0R1334-R1361)) to try to make this parameter handling a bit more readable. I think it's still more verbose than I'd like but clearer, what do you think?

Reply:
My first iteration did not fit with the helper function, nor one for `add`. But it's changed a lot since I last tried. I'll revisit it; readability and encapsulation is important.

Reply:
I did something similar to your suggest in [`49c8dcd` (#12662)](https://github.com/Qiskit/qiskit/pull/12662/commits/49c8dcd69f8c31dd0dd48be4425e9af6b3357dd4). This factors out cloning a `Param`. This is only used once now, but can be used whenever a `Param` needs to be cloned. I find the code much more readable at the call point.

Regarding factoring out `add_param`: There are pros and cons to doing this. Disadvantages: The branch with `-*phi + PI / 2.0` is more readable if it is not nested calls. In fact, this made it slightly easier for me to spot the error `-*phi + 1.0`.

Another one is that it requires two match statements rather than one, so `-*phi + PI / 2.0`  won't be optimized. I'm not sure, but I doubt this will matter in the future for performance.

But, I'm on the fence. Currently, the second match arm would be much more readable with your suggestion. I can implement and use `add_param` quickly if you prefer. I'm inclined to do it just to look at it.

Reply:
Implemented your suggestion in [`07fec37` (#12662)](https://github.com/Qiskit/qiskit/pull/12662/commits/07fec379319090b0b065b2589b63ddc75b7430c4). I'm not 100% sure there isn't a non-negligible efficiency hit. But I think it's easier to read and will be easier to build on.

Reply:
I see your point, I think we could merge the code as-is and once all Rust standard gates are in place, maybe run the benchmarks with/without `add_param` to see if there is indeed a significant performance hit.